Lib/gzip.py

```diff
@@ -245,8 +245,17 @@ def __init__(self, filename=None, mode=None,
 
         self.fileobj = fileobj
 
-        if self.mode == WRITE:
-            self._write_gzip_header(compresslevel)
+        try:
+            if self.mode == WRITE:
+                self._write_gzip_header(compresslevel)
+        except BaseException:
+            # Avoid a ResourceWarning if the write fails, eg read-only file or KeyboardInterrupt
+            try:
+                if self.myfileobj is not None:
+                    self.myfileobj.close()
+            finally:
+                self.fileobj = None
```

Comment:
What do you think of moving this code into a new _close() method, and modify close() to also call _close()?

Reply:
yeah that's much nicer, done!